packages/app/android/src/main/java/io/invertase/firebase/common/TaskExecutorService.java

```diff
@@ -0,0 +1,115 @@
+package io.invertase.firebase.common;
+
+/*
+ * Copyright (c) 2016-present Invertase Limited & Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this library except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+import io.invertase.firebase.common.ReactNativeFirebaseJSON;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.RejectedExecutionHandler;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.SynchronousQueue;
+
+public class TaskExecutorService {
+  private static final String MAXIMUM_POOL_SIZE_KEY = "android_task_executor_maximum_pool_size";
+  private static final String KEEP_ALIVE_SECONDS_KEY = "android_task_executor_keep_alive_seconds";
+
+  private final String name;
+  private final int maximumPoolSize;
+  private final int keepAliveSeconds;
+  private static Map<String, ExecutorService> executors = new HashMap<>();
+
+  TaskExecutorService(String name) {
+    this.name = name;
+    ReactNativeFirebaseJSON json = ReactNativeFirebaseJSON.getSharedInstance();
+    this.maximumPoolSize = json.getIntValue(MAXIMUM_POOL_SIZE_KEY, 1);
+    this.keepAliveSeconds = json.getIntValue(KEEP_ALIVE_SECONDS_KEY, 3);
+  }
+
+  public ExecutorService getExecutor() {
+    boolean isTransactional = maximumPoolSize <= 1;
+    return getExecutor(isTransactional, "");
+  }
+
+  public ExecutorService getTransactionalExecutor() {
+    return getExecutor(true, "");
+  }
+
+  public ExecutorService getTransactionalExecutor(String identifier) {
+    String executorIdentifier = maximumPoolSize != 0 ? identifier : "";
+    return getExecutor(true, executorIdentifier);
+  }
+
+  public ExecutorService getExecutor(boolean isTransactional, String identifier) {
+    String executorName = getExecutorName(isTransactional, identifier);
+    ExecutorService existingExecutor = executors.get(executorName);
+    if (existingExecutor != null) return existingExecutor;
+    ExecutorService newExecutor = getNewExecutor(isTransactional);
+    executors.put(executorName, newExecutor);
```

Comment:
I wonder if this should be synchronized on the executors object? Right now the method isn't synchronized and the state being mutated is the static set - is it possible for multiple getExecutor() calls to race and end up returning two different executors for use in calling code, even if the Set de-duplicates them?

I think by the nature of react-native being single-threaded it is not possible normally - but that may change in the future, and it means that lock contention at least should not be an issue


I recognize using a static object as a lock may represent a performance bottleneck of it's own but I consider that an implementation detail in that there are ways to optimize it but more importantly if it is required for correctness it must be done regardless

Reply:
I think you've got a fairly good point. As you mentioned, the multi-threading race issue might be hardly happened right now, but it also means that use of synchronized block hardly be a bottleneck. I see the synchronization with the executors object would eliminates the concern almost at no cost.